[Cache] memcache fix for J3.6.2, memcache(-d) performance improvements, less connections, add support for 'platform specific caching'

csthomas opened the pull request:
Pull Request for Issue (Not yet)
#### Summary

Cache storage memcache from J3.6.2 is not working.
This PR fix that.

I also add the same changes to memcached.

This is part of  another (more optimized) PR #10565. 
#### Summary of Changes

[UPDATED]

1) Fix compression issue.
Before only first instance of memcache can set compression to true on `getConnection` method.
Next instances did not call `getConnection` and then did not use compression at all.
Move `$this->_compress = ...` to consctructor.

https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aR47

2) Check `static::$_db === null` before `isSupported()`. 
If connection exists then skip call `isSupported` and do not create useless connections.
https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aL55

3) Fix persistent connection. Remove test connection which break persistent.
https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aL79

4) Remove race condition. We do not need to initiate empty list.
https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aL89

5) Joomla uses only one memcache server. Remove `replace` and use only `set` method.
https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aL194

6) Fix `remove` method. Search in all list instead of in first item only.
https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aL236

7) Do not test connection in `isSupported` method https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aL311
- In global configuration we should have an option to set another custom port if default does not work.

8) Do not lock twice. Do not call `lockindex` for `lock`/'unlock' method.
https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aL334
https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aL395

9) Fix `locklooped`.
https://github.com/joomla/joomla-cms/pull/11565/files#diff-8104780960bee59ef8a377ffd818797aR375

https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/cache/controller/callback.php#L99

10) Override `_getCacheId($id, $group)` for memcache(-d).
If `Platform specific caching` in `Global Configuration->Cache Settings` is enabled 
then mobile (user agent) user without that patch could not delete/see all cache data.
#### Testing Instructions

1) On Joomla 3.6.2 or newer try to use Cache handler MemcachE and MemcacheD
2) Test Joomla  with memcachE and MemcacheD handler on J 3.6.2 (without and with patch)
3) Check Backend->Cache Clear page.

(Optional)
Check how many connection (in memcached stats) generates joomla before patch and after patch.
After patch there should be less connections.
#### Other method to test.

https://github.com/joomla/joomla-cms/pull/11565#issuecomment-240566785

Commit:
Fix memcache storage, refresh code

photodude commented:
Looks like this PR introduces (or reveals) a memcache issue in HHVM 
`Fatal error: [] operator not supported for strings in /libraries/joomla/cache/storage/memcache.php on line 194`

csthomas commented:
I can change it but I think that HHVM do not use `unserialize` on `$index`.

`$index` all time should be `array` or `false` if not set.
Memcache implicitly serializes array to string and stores that string in memory. 
May be HHVM do not use unserialize after get variable and we have to do it explicitly.

photodude commented:
From what I can tell the following line is returning a string rather than an array. 

```php 
$index = static::$_db->get($this->_hash . '-index');
```

csthomas commented:
I can not confirm that on my local hhvm.

with hhvm 3.14.4 server from (deb http://dl.hhvm.com/ubuntu xenial main):

`hhvm -m server -d hhvm.server.port=9000 -d hhvm.server.source_root=/home/xxx/public_html/demo362/`

URLs:
* http://127.0.0.1:9000/administrator/index.php
* http://127.0.0.1:9000/

works.

I have got only:
"Warning: Division by zero in /home/tomash/public_html/demo362/plugins/system/debug/debug.php on line 589"

photodude commented:
I'm just going by the conditions necessary to reproduce the error as found on Travis CI. 

You can see the mock code at this link https://3v4l.org/HLGlZ

You can change $index to null or to false and there is no error. But if $index is a string the array modify opperator with throw the error.

csthomas commented:
IMHO there is bug in travis tests, or php-memcache is too old.
I use this one 3.0.9 from:
https://launchpad.net/ubuntu/xenial/+source/php-memcache

Example which I do not understand:
```
if [[ $INSTALL_MEMCACHE == "yes" ]]; then phpenv config-add build/travis/phpenv/memcached.ini; fi
```
Why there is memcacheD extension loaded instead memcachE?

csthomas commented:
Please correct me if I'm wrong.

Why did not travis see that memcachE can not connect to server after applied PR #9622.
Human tests missing memcachE too.

photodude commented:
This is telling Travis CI to adjust the PHP environment if the PHP version needs to make special settings for MEMCACHE 
```php
if [[ $INSTALL_MEMCACHE == "yes" ]]; then phpenv config-add build/travis/phpenv/memcached.ini; fi
```
It's specific to Travis and the container settings and only sets up MEMCACHE in the PHP environment. It's also not required for some PHP versions. In some versions both MEMCACHE and MEMCACHED are supported. 

Currently both MEMCACHE and MEMCACHED are supported, even though MEMCACHE is outdated and will go away at some future date. (HHVM only supports MEMCACHE and PHP 7 only supports MEMCACHED)
 I would agree there is a bug if the MEMCACHE connection wasn't being made and php unit Travis CI was still passing the unit tests.

photodude commented:
Looking deeper into what is happening at this line.

```php 
$index = static::$_db->get($this->_hash . '-index');
```

Memcache::get returns a string unless it was passed an array of keys.

http://php.net/manual/en/memcache.get.php

This is why $index is a string when the array modifier opperator tries to append $tmparr to the value.

csthomas commented:
> Currently both MEMCACHE and MEMCACHED are supported, even though MEMCACHE is outdated and will go away at some future date. (HHVM only supports MEMCACHE and PHP 7 only supports MEMCACHED)

I thought that too but when I checked yesterday MEMCACHE is still working on php7.0 - you can test it on ubuntu 16.04 LTS.

> if [[ $INSTALL_MEMCACHE == "yes" ]]; then phpenv config-add build/travis/phpenv/memcached.ini; fi

IMHO this is a bug/typo, it should be:
```
if [[ $INSTALL_MEMCACHE == "yes" ]]; then phpenv config-add build/travis/phpenv/memcache.ini; fi
```
I created a PR for that #11576.

mbabker commented on libraries/joomla/cache/storage/memcache.php:
```diff
@@ -31,15 +31,15 @@ class JCacheStorageMemcache extends JCacheStorage
 	 * @since  11.1
 	 */
-	protected $_persistent = false;
+	protected static $_persistent = false;
```
There's no need to make these static.  Actually I'd say that's a step backward.  The only reason the Memcache instance is static is because Joomla still doesn't believe in proper dependency injection but prefers global singletons.

csthomas replied:
Let me explain my point of view.

Before start set compression for memcache to true in Global configuration;

Example:

1) Component create instance of cache with some options like lifetime=600
2) Cache instance get instance of storage:
```
		$hash = md5(serialize($this->_options));
		if (isset(self::$_handler[$hash]))
		{
			return self::$_handler[$hash];
		}
		self::$_handler[$hash] = JCacheStorage::getInstance($this->_options['storage'], $this->_options);
```
3) first storage instance is created, compression=1, yes?
4) Next joomla get modules.
5) Module has different lifetime=900 then options is different then joomla create another instance
of storage.

6) Now we have to storages of memcache  - do you agree?.
7) First storage (before patch) has set compression=1 because getConnection() is called
8) Second storage instance (before patch) has compressio=0 because this does not need to create another connection, (because use existed static::$db)

What we have before patch? Only first instance of storage uses compression.

If you understand above, then
should I change it and each time of call __construct() set 
```php
$this->compression = JFactory::getConfig()->get('memcache_compress', false) ? MEMCACHE_COMPRESSED : 0;
```

Is it better?

mbabker replied:
This probably won't happen before 4.0 but I'd rather see all the adapters
have the internal connection injected versus us having such a static
singleton based API.  We should be able to create a single Memcache(d)
connection and drop that into both the cache and session as an example.
But back on track...

So right now because the connection is static there's little value in these
options being non static if they only affect the connection params.
However if they can affect other behaviors they should be configurable
per-instance.  Granted I don't see how this is actually useful with our
current structure but imagine a scenario where different cache data might
be stored on different Memcache(d) instances for whatever reason, each
connection should be configurable basically.

It just feels wrong that whatever ends up being the first call to these
class' constructors dictates the connection for the full request without
getting into some PHP magic (i.e. Reflection or overloading the autoloader
and creating custom classes).

On Saturday, August 13, 2016, Tomasz Narloch <notifications@github.com>
wrote:

> In libraries/joomla/cache/storage/memcache.php
> <https://github.com/joomla/joomla-cms/pull/11565#discussion_r74691810>:
>
> > @@ -31,15 +31,15 @@ class JCacheStorageMemcache extends JCacheStorage
> >  	 * @var    boolean
> >  	 * @since  11.1
> >  	 */
> > -	protected $_persistent = false;
> > +	protected static $_persistent = false;
>
> Let me explain my point of view.
>
> Before start set compression for memcache to true in Global configuration;
>
> Example:
>
> 1) Component create instance of cache with some options like lifetime=600
> 2) Cache instance get instance of storage:
>
>         $hash = md5(serialize($this->_options));
>         if (isset(self::$_handler[$hash]))
>         {
>             return self::$_handler[$hash];
>         }
>         self::$_handler[$hash] = JCacheStorage::getInstance($this->_options['storage'], $this->_options);
>
> 3) first storage instance is created, compression=1, yes?
> 4) Next joomla get modules.
> 5) Module has different lifetime=900 then options is different then joomla
> create another instance
> of storage.
>
> 6) Now we have to storages of memcache - do you agree?.
> 7) First storage (before patch) has set compression=1 because
> getConnection() is called
> 8) Second storage instance (before patch) has compressio=0 because this
> does not need to create another connection, (because use existed
> static::$db)
>
> What we have before patch? Only first instance of storage uses compression.
>
> If you understand above, then
> should I change it and each time of call __construct() set
>
> $this->compression = JFactory::getConfig()->get('memcache_compress', false) ? MEMCACHE_COMPRESSED : 0;
>
> Is it better?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11565/files/4241cbcde6f909b613017fdd4dac8ea9d1f64ad2#r74691810>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWfoUTZolV7LHqcIg9DyR2hsIfSwLGAks5qfihcgaJpZM4JiqOc>
> .
>

mbabker commented on libraries/joomla/cache/storage/memcache.php:
```diff
@@ -68,28 +68,36 @@ public function __construct($options = array())
-		return;
+		if (!$result)
+		{
+			throw new RuntimeException('Could not connect to memcache server');
```
`static::$_db` should be nulled back out here.  If it fails to connect, if you instantiate a second instance of this driver and the driver already has a static Memcache instance, that second instance will try to function as normal and won't throw an error.

Commit:
No static var

csthomas commented:
@mbabker I changed problematic stuff. 
I also removed `protected $_persistent = false;` because it is not using now. 
If it is B/C then I can back it, but it is useless.

photodude commented on libraries/joomla/cache/storage/memcache.php:
```diff
@@ -177,25 +177,17 @@ public function store($id, $group, $data)
-		{
-			$index = array();
-		}
+		$index = $index ? $index : array();
```
I'm not sure changing to a ternary operator is a good design choice here. The ternary operator in PHP does a copy before evaluating the statement, This results in a performance hit as a data set being evaluated grows (i.e. the more data in the $index variable the slower the ternary operator evaluation is)

Personally, I avoid using ternary operators with objects or arrays since you risk the performance hit from the variable copy it does before evaluation.

csthomas replied:
Thanks for tip.

photodude commented on libraries/joomla/cache/storage/memcache.php:
```diff
@@ -429,7 +357,7 @@ public function unlock($id, $group = null)
 	{
 		$looptime  = 300;
-		$data_lock = static::$_db->add($this->_hash . '-index_lock', 1, false, 30);
+		$data_lock = static::$_db->add($this->_hash . '-index_lock', 1, 0, 30);
```
I believe `false` is preferred over bool 0 or 1 in our code style. `false` is also used in the [`Memcache::add()` examples ](http://php.net/manual/en/memcache.add.php#refsect1-memcache.add-examples)

so possibly better as false.

mbabker replied:
Whatever the documented type is should be what's used.  If PHP 7 scalar
typehinting were applied and it fails, it should be changed.

On Sunday, August 14, 2016, Walt Sorensen <notifications@github.com> wrote:

> In libraries/joomla/cache/storage/memcache.php
> <https://github.com/joomla/joomla-cms/pull/11565#discussion_r74715747>:
>
> > @@ -429,7 +357,7 @@ public function unlock($id, $group = null)
> >  	protected function lockindex()
> >  	{
> >  		$looptime  = 300;
> > -		$data_lock = static::$_db->add($this->_hash . '-index_lock', 1, false, 30);
> > +		$data_lock = static::$_db->add($this->_hash . '-index_lock', 1, 0, 30);
>
> I believe false is preferred over bool 0 or 1 in our code style. false is
> also used in the Memcache::add() examples
> <http://php.net/manual/en/memcache.add.php#refsect1-memcache.add-examples>
>
> so possibly better as false.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11565/files/3097947c737cb9739787a57ef99051cfa904e5f3#r74715747>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWfoSK_pnJyWLGnNk6NoWKRDDQRMEJGks5qf9APgaJpZM4JiqOc>
> .
>

csthomas replied:
Another example http://php.net/manual/en/memcache.set.php#example-4966.
And definition:
```php
bool Memcache::set ( string $key , mixed $var [, int $flag [, int $expire ]] )
```

IMHO false was used some time ago, now we should use 0.

Commit:
Revert ternary operator

csthomas commented:
I have removed ternary and rebased to the latest staging to resolve the conflict.
Summary of changes has been updated.

photodude commented:
@csthomas Would you change [line 190](https://github.com/csthomas/joomla-cms/blob/memcache_1/libraries/joomla/cache/storage/memcache.php#L190) in the [`store()` method](https://github.com/csthomas/joomla-cms/blob/memcache_1/libraries/joomla/cache/storage/memcache.php#L158-L198) from from $index[] = $tmparr; to array_push($index, $tmparr); ? This solves [the HHVM error `Fatal error: [] operator not supported for strings in /libraries/joomla/cache/storage/memcache.php on line 190`](https://travis-ci.org/joomla/joomla-cms/jobs/152511380#L1121) that occurs when `$index = static::$_db->get($this->_hash . '-index');` returns a string. (or otherwise deal with the case when Memcache::get returns a string so we don't have this fatal error)

Commit:
Add test whether a $index is an array

csthomas commented:
Is the more issue there? 

Can I ask for test. 

Test should be simple: check whether cache works or not, test clear cache, etc.
For more advanced testers: code review.

mbabker commented:
> Test should be simple: check whether cache works or not, test clear cache, etc.

Finding people with access to certain PHP configurations is challenging at times.  I have MemcacheD set up on my local system but not Memcache and I don't have any servers running the latter to test this against.

csthomas commented:
If you have access to root account you could run: `pecl install memcache-3.0.8` for php5.
After test you can display installed items by `pecl list` and uninstall by `pecl uninstall memcache`.

csthomas commented:
May be someone with HHVM can test, memcache extension is probably built in.

csthomas commented:
Example which I use for hhvm on kubuntu 16.04 LTS.

1. Install hhvm
`$ sudo apt-get install hhvm`

2. Run www server on port 8000 and set source_root to joomla directory
`$ hhvm -m server -d hhvm.server.port=8000 -d hhvm.server.source_root=/home/[...]/public_html/demo362/`

3. Go to http://localhost:8000/administrator/index.php and test:)

csthomas commented:
@brianteeman Can I ask you to change priority to higher?  Currently cache handler MEMCACHE on J3.6.2 does not work at all.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11565">issues.joomla.org/joomla-cms/11565</a>.</sub>

photodude commented:
There is no current need for users to test against HHVM for code acceptance, same with php7.1. these are allowed failures in the unit tests while we make necessary code changes to add support. 

Code changes for HHVM and PHP 7.1 just need to check that they do not break existing supported PHP versions.

csthomas commented:
@photodude I know that but HHVM has memcache built in. 

May be it will be easier for testers to test on hhvm (if I can assume: if pass on hhvm then should also pass on php).

mbabker commented:
> If you have access to root account you could run

Everything I've got deployed is on PHP 7, Memcache isn't compatible with it.  I'd have to spend some time fighting with Homebrew and my Mac's setup to get Memcache installed on an older PHP version.

csthomas commented:
> Everything I've got deployed is on PHP 7, Memcache isn't compatible with it.

https://www.howtoforge.com/tutorial/install-apache-with-php-and-mysql-on-ubuntu-16-04-lamp/
http://www.ubuntuupdates.org/package/core/xenial/universe/base/php-memcache
https://github.com/websupport-sk/pecl-memcache

I know the problem may be with compile new version. 
For now I only know easy way for Ubuntu 16.04 LTS.
`sudo apt-get install php-memcache` which works with php7.

photodude commented on libraries/joomla/cache/storage/memcache.php:
```diff
@@ -178,7 +178,7 @@ public function store($id, $group, $data)
 		$index = static::$_db->get($this->_hash . '-index');
 
-		if ($index === false)
+		if (!is_array($index))
```
This is not the right change here.

Consider what variable types `$index` can have from line 179 .
```php 
$index = static::$_db->get($this->_hash . '-index');
```

`static::$_db->get(...)` .i.e [`Memcache::get`](http://php.net/manual/en/memcache.get.php) It could have a string type if it was passed a string key, or it could have an array type if an array of keys was passed to the method, or it could return the exact value boolean `false` on failure.

IMO this change incorrectly assumes that if $index is just a string associated with a key that the value can be replaced with an empty array before appending new data to be stored. We can do that with false since a failure at a `get()` (empty value found) before `set()` doesn't affect storing new data since any valid errors are handled elsewhere.

My suggestion of using [`array_push()`](http://php.net/manual/en/function.array-push.php) over the [array modifier `[]` syntax](http://php.net/manual/en/language.types.array.php#language.types.array.syntax.modifying) does handle the $index is a string or an array without additional conditional checks. (although there is some overhead, so there is some small performance hit here)

If you want to use conditionals you would need something like the following and type cast the string case of $index
```php
		if ($index === false)
		{
			$index = array();
		}
		elseif (!is_array($index))
		{
			$index = (array) $index;
		}
```

mbabker commented:
That's the first I've seen that someone's even working on the memcache extension, let alone that there's supposedly a PHP 7 compatible build.  Good to know.

photodude commented:
> May be it will be easier for testers to test on hhvm (if I can assume: if pass on hhvm then should also pass on php).

Few have HHVM set ups, and I doubt the CMS would currently run on HHVM without issues (we still have a lot of unit test failures on HHVM), additionally HHVM does have deviations in some behaviors from PHP and there are php extensions that HHVM does not currently support, while other extensions do not replicate all functions. As such I wouldn't directly assume that if HHVM passes that PHP will too.

csthomas commented:
> static::$_db->get(...) .i.e Memcache::get It could have a string type if it was passed a string key, or it could have an array type if an array of keys was passed to the method, or it could return the exact value boolean false on failure.

Why do you think that
```php
$index = static::$_db->get($this->_hash . '-index');
```

may contains string (which does not come from previous error)
`$index` as string come from memcacheD storage which use other format of serialize. 
Then memcache internal `unserialize` may fail and return string like **'{a:0:{}}'**

The only way to get `$index`as string is error come from internal memcache `unserialize`.

$index usually look like:
```php
Array
(
    [0] => stdClass Object
        (
            [name] => d41d8cd98f00b204e9800998ecf8427e-cache-_testing-aa178765666f875a86e12a7590eb296e
            [size] => 8
        )

    [1] => stdClass Object
        (
            [name] => d41d8cd98f00b204e9800998ecf8427e-cache-_testing-1167a6c94988dd05be7ccc5faed80fa6
            [size] => 8
        )

)
``` 
If we allow to create structure like:
```php
Array
(
    [0] => {a:0:{}}

    [1] => stdClass Object
        (
            [name] => d41d8cd98f00b204e9800998ecf8427e-cache-_testing-aa178765666f875a86e12a7590eb296e
            [size] => 8
        )

    [2] => stdClass Object
        (
            [name] => d41d8cd98f00b204e9800998ecf8427e-cache-_testing-1167a6c94988dd05be7ccc5faed80fa6
            [size] => 8
        )

)
```
then we will get errors in other methods like in '$memcache->getAll()'

csthomas commented:
> static::$_db->get(...) .i.e Memcache::get It could have a string type if it was passed a string key, or it could have an array type if an array of keys was passed to the method, or it could return the exact value boolean false on failure.

I read some similar text somewhere but it means ($m - instance of memcache):
```php
$m->set('a', 'a1');
$m->set('b', 'b1');
$m->get(array('a', 'b')); // this return array
$m->get('b'); // this return string
```
but
```php
$m->set('x', array('a', 'b'));
$m->get('x'); // this return array
```

photodude commented:
> Why do you think that
```php
$index = static::$_db->get($this->_hash . '-index');
```
may contains string (which does not come from previous error)

As I have stated, it's by definition of the `Memcache::get()` method. Check the [function return values listed in the manual](http://php.net/manual/en/memcache.get.php#refsect1-memcache.get-returnvalues)

> **Returns the string associated with the key or an array** of found key-value pairs when key is an array. **Returns FALSE on failure**, key is not found or key is an empty array.

[memcache**D**::get](http://php.net/manual/en/memcached.get.php) has a different return value from memcache::get. Since we are specifically dealing with memcache it is the memcache definition and return values that have to be considered. (easy to get tripped up between the two closely related extensions)

photodude commented:
> ```php
$m->set('x', array('a', 'b'));
$m->get('x'); // this return array
```

It's probably not unexpected to see an array in this case, it's propapbly one of those undocumented things where the "string" associated with the string key is actually an array so the return is an array.

try something like
> ```php
$m->set('x', 'a');
$m->get('x'); // this should return a string
```

csthomas commented:
The magic sentence from http://php.net/manual/en/memcache.get.php is:
> **Memcache::get()** returns previously stored data if an item with such key exists on the server at this moment. 

Which indicate that `set` trigger serialize if data is not string and then `get` trigger `unserialize`.

But:
http://php.net/manual/en/memcache.set.php
> **Note:**
Remember that resource variables (i.e. file and connection descriptors) cannot be stored in the cache, because they cannot be adequately represented in serialized state.

csthomas commented:
Tutorial for lazy testers:)

1. Create demo on https://demo.joomla.org/
You have to create demo with email/password. Do not create demo with "Lanch Test Drive" because you may not have access to cpanel.

2. After that you should be at https://ua.siteground.com/my_account.htm

3. Click on button "Access cpanel"

4. On cpanel on the felt side is button "Flush cache". Click on it.

5. Then click on "Level 3: Memcached" and enable Memcached server "On"

6. Take a look at details of memcached server, PORT, remeber it.

7. Back to your new joomla site and go to backend.

8. login to administrator area (subdomain, password)

9. Go to System->System Information->Php Information and serach for text "memcache support".
If it is enabled go next.

10. Install https://github.com/joomla-extensions/patchtester/releases/download/3.0.0-alpha2/com_patchtester.zip

11. Go to Components->Joomla Patch Tester

12. Click on fetch data and wait a while

13. Search for this PR, means type 11565

14. Apply that patch

15. Go to System->Global Configuration->System and enable memcache, set PORT as you saw before in cpanel and Save.

16.  Now you are on new memcache handler. 

Warning: If you miss something then you may have problem because joomla do not allow you to change any configuration. Then you have to back to cpanel, open file manager and turn off cache in configuration.php file manually.

photodude commented:
@csthomas I think there maybe a documentation error in the [php `Memcache::get()` manual](http://php.net/manual/en/memcache.get.php)

Based on your example testing, I think it should say something like the following
>Memcache::get returns previously stored data of an item if such key exists on the server at this moment
Return Values 
Returns the value associated with the found key, or an array of found key-value pairs when key is an array. Returns FALSE on failure, key is not found or key is an empty array.

in other words

`Memcache::get(key)` | Return description | return Types
------------ | ------------- | -------------
string key - `Memcache::get('a1')` | value associated with the key | the type of the key-value i.e. a string, or array
array keys- `Memcache::get(array(a1,a2))` | array of found key-value pairs | array
not found/empty array - `Memcache::get(array())` | not found or other failure | boolean `false`

so `memcache::get` might return a string for a single key if the key value was a string.


So in our code: we need to know what the value of `$this->_hash . '-index'` is in the unit test when that key gets passed to `Memcache::get(...)`. If `$this->_hash . '-index'` exists and it's a single value which is a string, then $index will be a string when we try to use the Array modifier `[]` syntax at line 190

csthomas commented:
First I want to ask for testing.

About documentation of php memcache: there is a little mess, but this PR is not for that.

Ok, I explain a little - it will be helpful to understand my next PR #10565:
memcache/memcahed for get/set should work similar (return type).

```php
$m->get('a') => return false # because it does not exist or was set to false
$m->set('a', array()) => $m->get('a')     => return array();
$m->set('a', (object) array()) => $m->get('a')   => return (object) array();
$m->set('a', array((object) array())) => $m->get('a') => return array((object) array());
$m->set('a', 123) => $m->get('a')       => return 123;
$m->set('a', false) => $m->get('a')     => return false;

// but:

$m->delete('a');
$flags = 0;
$m->get('a', $flags) => return false; // and if flags === 0 this means that value was not found
$m->set('a', false);
$flags = 0;
$m->get('a', $flags) => return false; // and if flags !== 0 this means that value was set as false
```

Memcache/(-d) use serialize for all types which is not string.
If something can not be serialized then throw error.

This performance problem is big $index with contains list of objects.
Imagine 1000 items in cache. 
This is 1000 objects in array which have to be unserialize/serialize on any memcache get index and set index.

Commit:
Merge with #11659

csthomas commented:
> There was 1 error:
>
> 1) JCacheStorageMemcachedTest::testCacheHit
>
> Memcached::get(): could not read long value, too big

> /home/travis/build/joomla/joomla-cms/tests/unit/core/helper/helper.php:52

> /home/travis/build/joomla/joomla-cms/libraries/joomla/cache/storage/memcached.php:187

> /home/travis/build/joomla/joomla-cms/tests/unit/core/case/cache.php:91

This means that memcache left some data which memcached interpreted as big int.
More info: https://secure.php.net/manual/en/book.memcache.php#115666

mbabker commented:
So to help clean up that stuff let's fix the [TestCoreCache::teardown() method](https://github.com/joomla/joomla-cms/blob/staging/tests/unit/core/case/cache.php#L59) to completely clean the cache storage instead of just cleaning a single group.  For this, I'd say add a new method to `JCacheStorage`, a `clear()` method (consistent with PSR-6) that just deletes all the things.

csthomas commented:
`clear` and `clean`. Is not too much mess?
May be `flush` method?

mbabker commented:
I only suggest `clear` since that's what the caching PSR uses.  `flush` works too.

photodude commented:
Clearing and cleaning post test is exactly what I tried to implement broadly  in my pr for unit test memory management. https://github.com/joomla/joomla-cms/pull/10685

mbabker commented:
All you're doing there is unsetting the object.  There isn't a destructor in the classes that's flushing the cache so all it's doing is helping with the PHP process' memory use, not actually clearing the cache storage.

Commit:
Add flush method to storages

photodude commented:
If during our unit tests Memcached::get(): could not read long value, too big, resulting in left over data being in the cache, are we sure that a similar situation could not happen in a production environment?

Sure we can, and should, fix the unit tests; but IMO we should take reasonable steps to ensure that the same errors won't occur in production.

mbabker commented:
That's in part why the user has a way to clear the cache via the admin UI or if they're able to run CLI scripts we ship one to clear expired cache (could probably add another to just clear the cache completely).  We can't account for every possible error condition, i.e. when data fails to serialize and unserialize out of the cache store correctly, so some things will require manual intervention no matter what.  I think that scenario hits the manual intervention criteria.

Commit:
Replace 'replace' method by 'set' method

csthomas commented:
What do you thing about the last two commits? 
Could I add the same changes (from memcache) to memcached in this PR?

csthomas commented:
Travis passed:)

mbabker commented:
If it works, go for it.  I honestly don't know the systems well enough, I can just create very high level API interactions.

csthomas commented:
I'm curious, may be I should use `clear` method instead 'flush' and use it as a replacement for clean in J4.
Then set clean as deprecated and create `clear`:
```php
public function clear($group = null)
```

then clear won't be full B/C because there won't be mode="notgroup".
I have not seen any use of $mode='notgroup' except tests.

photodude commented:
@mbabker I'm referring to the change at line 181. I still don't think that solves the underlying issue (cache has left over junk) and possibly introduces new issues. The whole cache clearing thing in the unit tests are revealed in the issue caused by line 190 in hhvm

csthomas commented:
IMHO The best solution for cache storages would be use the same design as in database tests.
Means move create storage from `setUp` to `setUpBeforeClass` and then `$storage->flush()` on `tearDownAfterClass()` method.

Take a look at https://github.com/joomla/joomla-cms/blob/staging/tests/unit/core/case/database.php

Commit:
Sync memcahe changes to memcached

photodude commented:
@csthomas I have no disagreement with the proposed changes for the Unit Test cache storages. I think it's a good choice to use the same design as the database tests. 

My only disagreement is with the change from https://github.com/joomla/joomla-cms/pull/11565/commits/a255c63030dc5aebb053e37fc6ac36437fbe7b2d as noted in the line comment.

Commit:
Fix issue with clean when getPlatformPrefix is set

csthomas commented:
@photodude We have:
```php
$index = static::$_db->get($this->_hash . '-index');
```

then `$this->_hash . '-index'` is always string, right?

Now `$index` can be:
- false, means not exists
- array, means exists and it is OK, but (may contains invalid data in items)
- string, means that this is invalid data because we always use array value for key `$this->_hash . '-index'`.

Now if we get string, means invalid data, what then?

You suggest to change it to array and push string at the first place in array. 
This off course trigger warning in `array_push` but who cares.

Most important things is that `$index` is a list of objects, but now we will have string at key=0.
This will generate errors on `getAll` or `clean` method.

csthomas commented:
I think that is enough for that PR. 
Memcache and Memcached has been changed. 

Travis tested memcache and memcached and passed.

Please test at least one storage.

photodude commented:
If your confident that when MEMCACHE::get is a string it is always invalid data and can be ignored in this case, then I'm fine with that change.

csthomas commented:
Yes but only for that key=`$this->_hash . '-index'`.
For other keys it may be string as valid data.

photodude commented:
I have tested this item :white_check_mark: successfully on 3aa94f581dc6e5a191c5b8ddb561a0476344bc7c<br /><br />Works, with this patch Memcache is now a selectable option for cache. Where before memcache only worked for the session handler.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11565">issues.joomla.org/joomla-cms/11565</a>.</sub>

yvesh commented on libraries/joomla/cache/storage/memcached.php:
```diff
@@ -68,34 +62,79 @@ public function __construct($options = array())
+
+			if ($servers)
+			{
+				if ($servers[0]['host'] != $host || $servers[0]['host'] != $port)
```
You can combine that line with the if one above (&&) and avoid one more level of nesting.

Also is this correct `$servers[0]['host'] != $port` ?

Commit:
Fix typo, avoid netsted if

csthomas commented:
Thanks @yvesh for code review.
Thanks @photodude for your test.

After my commit I probably still need 2 success tests.

Commit:
Next typo in msg Memcache => Memcached

gwsdesk commented:
Maybe a stupid question but we have Memcached running on our servers and on our Joomla  3.6.2  websites without any problems (PHP 5.6.24 with cloudlinux)   So what do we need to fix which works for us?

csthomas commented:
> Maybe a stupid question but we have Memcached

This PR only add improvement for memcacheD

But you can not use MemcachE on J3.6.2

csthomas commented:
If `memcache` handler does not work on J3.6.2 and nobody complain about it (except me) then could testers test at least `memcached` handler.

There are the same changes in both, except `getConnection()`, which has a few more differences.

@photodude @mbabker Can I ask you to to make a test?

gwsdesk commented:
I have tested Memcache**d** :white_check_mark: on PHP 5.6.24/Cloudlinux/Mariadb/cgi-fcgi on Joomla 3.6.2 successfully This runs well on a Joomla 3.6.2 multi lingual testsite gwsdev2.work

Short remark on Memcache. Memcache is not available yet for PHP7. This is the reason why we have removed it from the PHP-extensions in the PHP-selector of Cloudlinux altogether and for now just provide Memcache**d**. This is not any issue since in the CL environment you get OPCache and that works just so nice and fast in combination with Memcache**d**

csthomas commented:
Thank you @gwsdesk.

> Short remark on Memcache. Memcache is not available yet for PHP7

Official not, but at least one linux distribution have memcache working with PHP7.
 
If you have Ubuntu 16.04 on some machine then you can install php-memcache (PHP7) from default repository. There is probably unofficial version of memcache compiled in deb.

https://github.com/joomla/joomla-cms/pull/11565#issuecomment-240481391

gwsdesk commented:
Thanks Thomasz but Cloudlinux does not support Memcache in PHP7 yet. In 
Centos we can also install Memcache from default and that works I can 
confirm. CL no go at present though they have a Beta version


On 8/25/2016 3:33 PM, Tomasz Narloch wrote:
>
> Thank you @gwsdesk <https://github.com/gwsdesk>.
>
>     Short remark on Memcache. Memcache is not available yet for PHP7
>
> Official not, but at least one linux distribution have memcache 
> working with PHP7.
>
> If you have Ubuntu 16.04 on some machine then you can install 
> php-memcache (PHP7) from default repository. There is probably 
> unofficial version of memcache compiled in deb.
>
> #11565 (comment) 
> <https://github.com/joomla/joomla-cms/pull/11565#issuecomment-240481391>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/joomla/joomla-cms/pull/11565#issuecomment-242316637>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AHzLNkC-weu902hW0b08oA4bdnuGK06zks5qjVNfgaJpZM4JiqOc>.
>

csthomas commented:
Can your test can be mark as successful at https://issues.joomla.org/tracker/joomla-cms/11565

If you have a few minutes You can test memcache as follow:
https://github.com/joomla/joomla-cms/pull/11565#issuecomment-240566785

gwsdesk commented:
I have tested this item :white_check_mark: successfully on 4e5e96be91e17ff2b75a074ed06f0672332114a5<br /><br />I have tested Memcache**d** on PHP 5.6.24/Cloudlinux/Mariadb/cgi-fcgi on Joomla 3.6.2 successfully :white_check_mark: .This runs well on a Joomla 3.6.2 multi lingual testsite gwsdev2.work. I also tested this on the same site with PHP7 enabled and works fine as well<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11565">issues.joomla.org/joomla-cms/11565</a>.</sub>

photodude commented:
I have tested this item :white_check_mark: successfully on 4e5e96be91e17ff2b75a074ed06f0672332114a5<br /><br />
Marking successful, with the following note.


As soon as I applied this revised patch I got the following notice

`Warning: Error loading component: com_languages, Could not connect to memcache server`

I assume this warning was due to an old cache state; as it went away after turning cache off and then back on in Global settings. Since it was a transient message and resolved after resetting the cache I don't think there is anything to do but note the message.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11565">issues.joomla.org/joomla-cms/11565</a>.</sub>

gwsdesk commented:
@photodude this means that it is not successful since it shows the original error (which goes away if you disable caching or change to file. Please remove the 'successful' tag and replace with 'unsuccessful' ? :x:
This is not just a 'warning' is means it does not work!

csthomas commented:
I don not know @photodude steps but I assume he enabled memcachE before applying patch.
Warning was stored in session and displayed at next page view.

Besides above. 
There was not any changes added to memcachE handler from the first @photodude test.
The last 2 commits only added changes to memcacheD handler.

photodude commented:
Here is what I had done. @gwsdesk
1. tested the patch in its old state using patch tester (see first test)
2. Removed the patch
3. Applied the patch in its current state
4. Warning appeared at this point
5. Turned off cache in global settings (forced logout)
6. (Logged in) turned cache back on 
7. No warnings or issues using joomla front end or back end. 

As you can see the warning was due to the inconsistent state the cache was in between reapplying this patch after it was updated. I stand by my choice in marking it as a successful test.

gwsdesk commented:
@photodude I read it correctly now and only noticed now your initial test as well. Thanks for successful testing ;-)

photodude commented:
@rdeutz can this be marked RTC now?

rdeutz commented:
@photodude yes it can<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11565">issues.joomla.org/joomla-cms/11565</a>.</sub>

zero-24 commented:
@joomla-cms-bot can you please help here? Thanks :smile:
